dedupe/datamodel.py

```diff
@@ -183,30 +178,33 @@ def typify_variables(
         try:
             variable_class = VARIABLE_CLASSES[variable_type]
         except KeyError:
+            valid = ", ".join(VARIABLE_CLASSES)
             raise KeyError(
-                "Field type %s not valid. Valid types include %s"
-                % (definition["type"], ", ".join(VARIABLE_CLASSES))
+                f"Variable type {variable_type} not valid. Valid types include {valid}"
             )
-
         variable_object = variable_class(definition)
-        assert isinstance(variable_object, FieldVariable)
-
-        primary_variables.append(variable_object)
-
-        if hasattr(variable_object, "higher_vars"):
-            all_variables.extend(variable_object.higher_vars)
-        else:
-            variable_object = cast(Variable, variable_object)
-            all_variables.append(variable_object)
+        assert isinstance(variable_object, Variable)
+        variables.append(variable_object)
 
+    only_custom = all(isinstance(v, (CustomType, InteractionType)) for v in variables)
     if only_custom:
         raise ValueError(
             "At least one of the variable types needs to be a type"
             "other than 'Custom'. 'Custom' types have no associated"
             "blocking rules"
         )
 
-    return primary_variables, all_variables
+    return variables
+
+
+def _expand_higher_variables(variables: Iterable[Variable]) -> list[Variable]:
```

Comment:
we could probably do away with this method if we had the base variable type have a higher_vars attribute set to an empty list or tuple

Reply:
no that won't quite work. it still feels like some of this should maybe be the responsible for the variable class.

Reply:
It definitely feels gross, there needs to be some more official API. Trying not to deal with that here.